### PR TITLE
bcc/tools: use device number and inode number to identify a file

### PR DESCRIPTION
**Pull request by chenhengqi**
Currently, the filetop tool use (tid, filename, type) tuple to
key a file, which is not enough to uniquely identify a file.
A thread write to multi files with the same name would add up to
same value in the map which can be repro by the following command:

$ cat somefile | tee /foo/bar/xxx /fuz/baz/xxx

Let us add device number and inode number to uniquely identify
a file.

Signed-off-by: Hengqi Chen <chenhengqi@outlook.com>

### CODE REVIEW

**Comment by chenhengqi** [comment]
Thanks @davemarchevsky  for pointing this out.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by yonghong-song** [comment]
[buildbot, ok to test]

**Review by anakryiko — COMMENTED** [review]
Does this work with btrfs? I remember we had some issues with btrfs due to some inode or dev reporting peculiarities.

**Comment by davemarchevsky** [comment]
@anakryiko I think you're referring to [this kludge](https://github.com/iovisor/bcc/blob/722cf83941879c52ebea5e5a1692b2976de6ad62/src/cc/api/BPF.h#L368-L382) and the various PRs and conversations around it ([1](https://github.com/iovisor/bcc/pull/2324) [2](https://github.com/iovisor/bcc/issues/2715) [3](https://github.com/iovisor/bcc/pull/2723) - just listing for completeness).

TL;DR of the issue is 

> btrfs and overlayfs report different device numbers for files in /proc/PID/maps vs stat syscall

so if we're getting device numbers from different sources and expecting to be able to compare across them it could be a problem e.g. common case of "pull out inode,dev from using BPF and find matching `/proc/PID/maps` entry". Since this tool is just using the device nums from one source it won't cause problem here.

**Comment by davemarchevsky** [comment]
LGTM, test failure only occurs on `fc27` and is unrelated to this PR

**Comment by anakryiko** [comment]
> @anakryiko I think you're referring to [this kludge](https://github.com/iovisor/bcc/blob/722cf83941879c52ebea5e5a1692b2976de6ad62/src/cc/api/BPF.h#L368-L382) and the various PRs and conversations around it ([1](https://github.com/iovisor/bcc/pull/2324) [2](https://github.com/iovisor/bcc/issues/2715) [3](https://github.com/iovisor/bcc/pull/2723) - just listing for completeness).
> 
> TL;DR of the issue is
> 
> > btrfs and overlayfs report different device numbers for files in /proc/PID/maps vs stat syscall
> 
> so if we're getting device numbers from different sources and expecting to be able to compare across them it could be a problem e.g. common case of "pull out inode,dev from using BPF and find matching `/proc/PID/maps` entry". Since this tool is just using the device nums from one source it won't cause problem here.

Ah, I see, thanks for explaining!

**Comment by yonghong-song** [comment]
[buildbot, test this please]
